src/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -1091,7 +1098,7 @@ private HeaderDescriptor GetHeaderDescriptor(string name)
                 return descriptor.AsCustomHeader();
             }
 
-            throw new InvalidOperationException(SR.net_http_headers_not_allowed_header_name);
+            throw new InvalidOperationException(string.Format(SR.net_http_headers_not_allowed_header_name, name));
         }
 
         private bool TryGetHeaderDescriptor(string name, out HeaderDescriptor descriptor)
```

Comment:
Due to the `!= 0`, these two lines will unintentionally test true if `descriptor.HeaderType` and `_allowHeaderTypes` containing `NonTrailing`.

https://github.com/dotnet/corefx/blob/7eca2f0f6c51ff517657c943d602cca4b81fddf4/src/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs#L1092

https://github.com/dotnet/corefx/blob/7eca2f0f6c51ff517657c943d602cca4b81fddf4/src/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs#L1117

Reply:
Valid point and to fix it I had to introduce more bit-wise ```&``` operators. In example, the code above now looks like 
```if (((descriptor.HeaderType & ~HttpHeaderType.NonTrailing) & _allowedHeaderTypes) != 0)```
where ```_allowedHeaderTypes``` is initialized as follows
```_allowedHeaderTypes = allowedHeaderTypes & ~HttpHeaderType.NonTrailing```
It seems really awkward, so I'd like to again raise my question on whether we really want to go with NonTrailing flag approach. Is it really worth the performance gain we get here? Trailing headers are rare.

Reply:
>Is it really worth the performance gain we get here? Trailing headers are rare.

@stephentoub already [commented](https://github.com/dotnet/corefx/pull/42229#discussion_r341194878) that we should prioritize performance.

I understand that some of the code might look akward due to the bit masking operations. But I previously suggested to create some helper methods that appropriately annotated would be inline-able by the compiler. That would make the code look cleaner.

Basically, the design change is here is moving from an enumeration describing the HeaderType to a bitmask describing the HeaderType. And bitmasks require various  '|' and '&' operations at times. Using helper methods to make the code look cleaner would help here.

Reply:
In the given cases helper methods are not sufficient because the comparison works only when _allowedHeaderTypes and _treatAsCustomHeaderTypes are properly initialized. Of course, instead of the one time initialization we can repeat the operation ```& ~HttpHeaderType.NonTrailing``` for them as well in the helper method, but anyways it looks unnatural.
My point here that we are trying to optimize **a rare case** by trading it off for code cleanness and thus **increasing chance of bugs** in the future. I would agree if trailing headers were on the hot path, then it would be justified. However, it seems they are not.

Reply:
@stephentoub can you comment on this? If we think the performance benefit is not that great then maybe the bitflag approach is not useful.

Reply:
If _`allowedHeaderTypes` is made to not contain `NonTrailing`, then the existing line `if ((descriptor.HeaderType & _allowedHeaderTypes) != 0)` does not need to be changed. Other comparisons, instead of having `if((headerType & foo) == foo)` can be made into `if((headerType & foo) != 0)` as well to avoid masking out `NonTrailing` from `headerType`.

Reply:
@scalablecory Sorry, I didn't get your idea completely. Could you please explain how the proposed simplification can be applied to HttpHeaders.GetHeaderDescriptor and HttpHeaders.TryGetHeaderDescriptor methods?
In these 2 methods, checks must pass even when the given descriptor.HeaderType has NonTrailing flag. In example, if I just use ```if ((descriptor.HeaderType & _allowedHeaderTypes) != 0)``` the check will **fail for the Host** header which has **Request and NonTrailing flags**.
[In the current PR version](https://github.com/dotnet/corefx/blob/7d5a59fd67709575aa823b3ecb59a8b4b51f8412/src/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs#L1092) I use ```if (((descriptor.HeaderType & ~HttpHeaderType.NonTrailing) & _allowedHeaderTypes) != 0)``` where[ _allowedHeaderTypes is initialized](https://github.com/dotnet/corefx/blob/7d5a59fd67709575aa823b3ecb59a8b4b51f8412/src/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs#L57) as
```_allowedHeaderTypes = allowedHeaderTypes & ~HttpHeaderType.NonTrailing```. 

Could you please show how it will look in your version?

Reply:
> we are trying to optimize a *rare case*

Lets not get hung up on an optimization red herring. Bit flags being an optimization is just a value-add, not the goal. The issue here is that the PR was needlessly triplicating a collection.

Bit flags are a suggestion that can cleanly solve that, but I am not adamant that we use them. If we can avoid triplicating the collection in another way that we think is clearer and easier to use without impacting perf, I am fine with that.

> if trailing headers were on the hot path, then it would be justified.

Trailing headers are used heavily in gRPC, and that team has communicated a want for optimization from us. Header parsing is the single most expensive part of our code other than perhaps TLS, so we should make sure to look at it with a keen eye.

> increasing chance of bugs in the future.

Bit manipulation is fairly common in protocol work, so I don't think we should shy away from it. In my experience it is not particularly bug-prone, but we should by all means do what we can to make us more successful iterating this code in the future. Renaming `HttpHeaderType` to `HttpHeaderFlags` and implementing helper methods where appropriate as @davidsh suggests are two things worth exploring.

Reply:
> Could you please show how it will look in your version?

So the bug is that `if((descriptor.HeaderType & _allowedHeaderTypes) != 0)` decomposed into e.g.:

```c#
descriptor.HeaderType = Content | NonTrailing;
_allowedHeaderTypes = Request | NonTrailing;

int mask = descriptor.HeaderType & _allowedHeaderTypes; // mask is now NonTrailing.
if(mask != 0)
{
   // invalid true path: we wanted Request, not Content, but mask is != 0 due to testing for NonTrailing.
}
```

To fix this, we only need to remove `NonTrailing` from one side of that `&` expression. The change you made to `_allowedHeaderTypes` resolves the problem without making any other code changes:
```c#
_allowedHeaderTypes = allowedHeaderTypes & ~HttpHeaderType.NonTrailing;
```

Now that `if` decomposes into:

```c#
descriptor.HeaderType = Content | NonTrailing;
_allowedHeaderTypes = Request; // no longer testing for NonTrailing.

int mask = descriptor.HeaderType & _allowedHeaderTypes; // mask is now 0.
if(mask != 0)
{
   ...
}
```